### PR TITLE
Enforce stability of const fn in promoteds

### LINKED ISSUE #50901: rustc_const_unstable attribute ignored in const promotion
```rust
let x: &'static _ = &std::time::Duration::from_millis(42).subsec_millis();
```

works on nightly but not on beta, because the `subsec_millis` function wasn't unstably const fn when beta was split off.

### PR DESCRIPTION
**Pull request by oli-obk**
r? @eddyb 

fixes #50901

what's going on here? Why do we have two promoted computation algorithms?

### CODE REVIEW

File: src/librustc_mir/transform/qualify_consts.rs [hunk 1]
```diff
@@ -975,14 +972,24 @@ This does not pose a problem by itself because they can't be accessed directly."
...
-                        help!(&mut err,
-                              "in Nightly builds, add `#![feature({})]` \
-                               to the crate attributes to enable",
-                              feature_name);
-                        err.emit();
+                        if self.mode == Mode::Fn {
+                            // We are in a normal function
+                            // with a turned off feature gate. We can still call the function
+                            // but we can't promote it
+                            self.qualif = Qualif::NOT_CONST;
```
Comment (eddyb) [inline src/librustc_mir/transform/qualify_consts.rs]: Why is this inside the `if`?  IMO the flags should be as general as possible, so we can make sure less problems show up again. Also, shouldn't this be something like `self.add(Qualif::NOT_CONST)`?

Comment (oli-obk) [reply src/librustc_mir/transform/qualify_consts.rs]: Doesn't make much difference as the other branch errors, but seems more consistent indeed

**Comment by eddyb** [comment]
@bors try

(we'll likely need a check-only crater run on this)

**Comment by oli-obk** [comment]
In case of regressions we could just stabilize the relevant const fns together with fixing the stability hole

**Comment by oli-obk** [comment]
Ping @rust-lang/infra can you start a check-only crater run?

**Comment by pietroalbini** [comment]
Crater run (check-only) started.

**Comment by pietroalbini** [comment]
Hi @oli-obk (crater requester), @eddyb (PR reviewer)! Crater results are at: http://cargobomb-reports.s3.amazonaws.com/pr-50909/index.html. 'Blacklisted' crates (spurious failures etc) can be found [here](https://github.com/rust-lang-nursery/crater/blob/master/config.toml). If you see any spurious failures not on the list, please make a PR against that file.

(interested observers: Crater is a tool for testing the impact of changes on the crates.io ecosystem. You can find out more at the [repo](https://github.com/rust-lang-nursery/crater/) if you're curious)

**Comment by eddyb** [comment]
Looks clean! (spurious regression)

r=me after rebase

**Pull request merged**